hibernate-infinispan/src/main/java/org/hibernate/cache/infinispan/access/TombstoneCallInterceptor.java

```diff
@@ -0,0 +1,184 @@
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.cache.infinispan.access;
+
+import org.hibernate.cache.infinispan.util.FutureUpdate;
+import org.hibernate.cache.infinispan.util.TombstoneUpdate;
+import org.hibernate.cache.infinispan.util.Tombstone;
+import org.infinispan.AdvancedCache;
+import org.infinispan.commands.read.SizeCommand;
+import org.infinispan.commands.write.PutKeyValueCommand;
+import org.infinispan.commons.util.CloseableIterable;
+import org.infinispan.container.entries.CacheEntry;
+import org.infinispan.container.entries.MVCCEntry;
+import org.infinispan.context.Flag;
+import org.infinispan.context.InvocationContext;
+import org.infinispan.factories.annotations.Inject;
+import org.infinispan.interceptors.CallInterceptor;
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+
+/**
+ * Note that this does not implement all commands, only those appropriate for {@link TombstoneAccessDelegate}
+ * and {@link org.hibernate.cache.infinispan.impl.BaseTransactionalDataRegion}
+ *
+ * The behaviour here also breaks notifications, which are not used for 2LC caches.
+ *
+ * @author Radim Vansa &lt;rvansa@redhat.com&gt;
+ */
+public class TombstoneCallInterceptor extends CallInterceptor {
+	private static final Log log = LogFactory.getLog( TombstoneCallInterceptor.class );
+	private static final UUID ZERO = new UUID(0, 0);
+
+	private AdvancedCache cache;
+
+	@Inject
+	public void injectDependencies(AdvancedCache cache) {
+		this.cache = cache;
+	}
+
+	@Override
+	public Object visitPutKeyValueCommand(InvocationContext ctx, PutKeyValueCommand command) throws Throwable {
+		MVCCEntry e = (MVCCEntry) ctx.lookupEntry(command.getKey());
+		if (e == null) {
+			return null;
+		}
+		Object value = command.getValue();
+		if (value instanceof TombstoneUpdate) {
+			return handleTombstoneUpdate(e, (TombstoneUpdate) value);
+		}
+		else if (value instanceof Tombstone) {
+			return handleTombstone(e, (Tombstone) value);
+		}
+		else if (value instanceof FutureUpdate) {
+			return handleFutureUpdate(e, (FutureUpdate) value);
+		}
+		else {
+			return super.visitPutKeyValueCommand(ctx, command);
+		}
+	}
+
+	private Object handleFutureUpdate(MVCCEntry e, FutureUpdate futureUpdate) {
+		Object storedValue = e.getValue();
+		if (storedValue instanceof FutureUpdate) {
+			FutureUpdate storedFutureUpdate = (FutureUpdate) storedValue;
+			if (futureUpdate.getUuid().equals(storedFutureUpdate.getUuid())) {
+				if (futureUpdate.getValue() != null) {
+					// transaction succeeded
+					setValue(e, futureUpdate.getValue());
+				}
+				else {
+					// transaction failed
+					setValue(e, storedFutureUpdate.getValue());
+				}
+			}
+			else {
+				// two conflicting updates
+				setValue(e, new FutureUpdate(ZERO, null));
```

Comment:
Hmm, if the current `FutureUpdate` is pre-, then yes, since the metadata is always set from the last put. But I forgot to add the expiration to the post-, good point. Btw., if you'd like to propose a better name, go on.